device/sycl/src/clusterization/measurement_creation.sycl

```diff
@@ -0,0 +1,138 @@
+/** TRACCC library, part of the ACTS project (R&D line)
+ *
+ * (c) 2022 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+// SYCL library include(s)
+#include "../utils/get_queue.hpp"
+#include "traccc/sycl/clusterization/measurement_creation.hpp"
+
+// SYCL include(s)
+#include <CL/sycl.hpp>
+
+// Vecmem include(s)
+#include <vecmem/utils/sycl/copy.hpp>
+
+namespace traccc::sycl {
+
+measurement_creation::measurement_creation(vecmem::memory_resource &mr,
+                                           queue_wrapper queue)
+    : m_mr(mr), m_queue(queue) {}
+
+host_measurement_container measurement_creation::operator()(
+    const host_cluster_container &c, const std::vector<std::size_t> &s,
+    const host_cell_module_collection &l) const {
+    output_type measurements(&m_mr.get());
+    this->operator()(c, s, l, measurements);
+    return measurements;
+}
+
+void measurement_creation::operator()(
+    const host_cluster_container &clusters,
+    const std::vector<std::size_t> &cluster_sizes,
+    const host_cell_module_collection &cell_modules_per_event,
+    output_type &measurements) const {
+
+    vecmem::sycl::copy copy{m_queue.queue()};
+    unsigned int size_buf = cell_modules_per_event.size();
+
+    // Resizable buffer for the measurements
+    measurement_container_buffer measurement_buffer{
+        {size_buf, m_mr.get()},
+        {std::vector<std::size_t>(cluster_sizes.size(), 0), cluster_sizes,
+         m_mr.get()}};
+    copy.setup(measurement_buffer.items);
+
+    // Vector views to move data to the device
+    auto clusters_data = get_data(clusters, &m_mr.get());
+    cluster_container_const_view clusters_view(clusters_data);
+    measurement_container_view measurement_view(measurement_buffer);
+
+    // range of kernel execution
+    auto range = clusters.size();
+
+    // Run the kernel
+    details::get_queue(m_queue)
+        .parallel_for(
+            range,
+            [clusters_view, measurement_view](auto idx) {
+                // Initialize device vectors
+                device_cluster_const_container clusters_device(clusters_view);
+                device_measurement_container measurement_device(
+                    measurement_view);
+
+                // Ignore if idx is out of range
+                if (idx >= clusters_device.size())
+                    return;
+
+                // items: cluster of cells at current idx
+                // header: cluster_id object with the information about the cell
+                // module
+                const auto &cluster = clusters_device.get_items().at(idx);
+                const auto &cluster_id = clusters_device.get_headers().at(idx);
+
+                const auto pitch = get_pitch(cluster_id.pixel);
+                const auto module_idx = cluster_id.module_idx;
```

Comment:
There is way too much `auto` being used here for my taste. :frowning: Please use explicit types, otherwise the lay reader will have no idea what is going on in this code...